server/src/middleware/rpc/mod.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any
+// person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the
+// Software without restriction, including without
+// limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software
+// is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Various middleware implementations for JSON-RPC specific purposes.
+
+mod layers;
+
+pub use layers::*;
+
+use std::net::SocketAddr;
+
+use http::{HeaderMap, Uri};
+use tower::layer::util::{Identity, Stack};
+use tower::layer::LayerFn;
+use tower::util::Either;
+
+use jsonrpsee_core::server::MethodResponse;
+use jsonrpsee_types::Request;
+
+/// Similar to `tower::Service` but specific for jsonrpsee and
+/// doesn't requires `&mut self` for performance reasons.
+#[async_trait::async_trait]
+pub trait RpcServiceT<'a> {
+	/// Process a single JSON-RPC call it may be a subscription or regular call.
+	/// In this interface they are treated in the same way but it's possible to
+	/// distinguish those based on the `MethodResponse`.
+	async fn call(&self, request: Request<'a>, transport_label: TransportProtocol) -> MethodResponse;
+}
+
+/// The transport protocol used to send or receive a call or request.
+#[derive(Debug, Copy, Clone)]
+pub enum TransportProtocol {
+	/// HTTP transport.
+	Http,
+	/// WebSocket transport.
+	WebSocket,
+}
+
+impl std::fmt::Display for TransportProtocol {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		let s = match self {
+			Self::Http => "http",
+			Self::WebSocket => "websocket",
+		};
+
+		write!(f, "{s}")
+	}
+}
+
+/// Context to a JSON-RPC call.
+#[derive(Debug, Clone)]
+pub struct Context {
+	/// Transport protocol.
+	pub transport: TransportProtocol,
+	/// Remote addr.
+	pub remote_addr: SocketAddr,
+	/// Connection id.
+	pub conn_id: usize,
+	/// HTTP Headers.
+	pub headers: HeaderMap,
+	/// URI.
+	pub uri: Uri,
+}
```

Comment:
Can this be deleted now?

Reply:
yes, hmm weird I thought I removed it :P